File: src/main/java/org/ffpy/plugin/coding/constant/CommentPosition.java
package org.ffpy.plugin.coding.constant;

import lombok.Getter;
import lombok.RequiredArgsConstructor;
import org.dom4j.Element;
import org.dom4j.Node;

import java.util.Collections;
import java.util.List;
import java.util.Optional;
import java.util.function.Function;

/**
 * 注释位置
 */
@RequiredArgsConstructor
public enum CommentPosition {
    UP("节点上方",
            e -> e.selectNodes("./preceding-sibling::comment()[1]"),
            n -> n.selectNodes("./following-sibling::*")),
    DOWN("节点下方",
            e -> e.selectNodes("./following-sibling::comment()[1]"),
            n -> n.selectNodes("./preceding-sibling::*")),
    NONE("无", e -> Collections.emptyList(), n -> Collections.emptyList()),
    ;
    @Getter
    private final String name;
    private final Function<Element, List<Node>> finder;
    private final Function<Node, List<Node>> prev;

    /**
     * 获取标签对应的注释节点
     *
     * @param el 标签
     * @return 注释节点
     */
    public Optional<Node> findComment(Element el) {
        return finder.apply(el).stream()
                .filter(node -> findNodeByComment(node) == el)
                .findFirst();
    }

    /**
     * 获取注释节点对应的标签
     *
     * @param node 注释节点
     * @return 标签
     */
    private Node findNodeByComment(Node node) {
        List<Node> list = prev.apply(node);
        if (list.isEmpty()) return null;
        return list.get(list.size() - 1);
    }
}

File: src/main/kotlin/org/ffpy/plugin/coding/constant/CommentPosition.kt
package org.ffpy.plugin.coding.constant

import org.dom4j.Element
import org.dom4j.Node

/**
 * 注释位置
 */
enum class CommentPosition(
    val text: String,
    private val findCommentByNodeAction: (Node) -> List<Node>,
    private val findNodeByCommentAction: (Node) -> List<Node>
) {
    UP("节点上方",
        { it.selectNodes("./preceding-sibling::comment()[1]") },
        { it.selectNodes("./following-sibling::*") }),

    DOWN("节点下方",
        { it.selectNodes("./following-sibling::comment()[1]") },
        { it.selectNodes("./preceding-sibling::*") }),

    NONE("无",
        { emptyList() },
        { emptyList() }),
    ;

    /**
     * 获取标签对应的注释节点
     *
     * @param el 标签
     * @return 注释节点
     */
    fun findComment(el: Element): Node? {
        return findCommentByNodeAction(el).firstOrNull()?.let {
            if (findNodeByComment(it) === el) it else null
        }
    }

    /**
     * 获取注释节点对应的标签
     *
     * @param node 注释节点
     * @return 标签
     */
    private fun findNodeByComment(node: Node): Node? {
        val list = findNodeByCommentAction(node)
        return if (list.isEmpty()) null else list[list.size - 1]
    }
}